distributed/active_memory_manager.py

```diff
@@ -0,0 +1,429 @@
+import asyncio
+import heapq
+import math
+from collections import defaultdict
+from collections.abc import Generator, Iterable, Iterator
+from typing import Optional
+
+import dask
+
+from .scheduler import SchedulerPlugin, SchedulerState, TaskState, WorkerState
+
+
+class ActiveMemoryManager(SchedulerPlugin):
+    scheduler: SchedulerState
+    plugins: "set[ActiveMemoryManagerPlugin]"
+    interval: float
+
+    # Daemon that runs every <interval> seconds
+    run_task: asyncio.Task
+
+    # These attributes only exist within the scope of self.run()
+    # Current memory (in bytes) allocated on each worker, plus/minus pending actions
+    workers_memory: "dict[WorkerState, int]"
+    # Pending replications and deletions for each task
+    pending: "defaultdict[TaskState, tuple[set[WorkerState], set[WorkerState]]]"
+
+    def __init__(
+        self,
+        plugins: "Iterable[ActiveMemoryManagerPlugin]" = None,
+        interval: float = 2.0,
+    ):
+        self.interval = interval
+        self.plugins = set()
+        if plugins is None:
+            plugins = {ReplicateReduce(), Rebalance()}
+        for plugin in plugins:
+            self.register_plugin(plugin)
+
+    def register_plugin(self, plugin: "ActiveMemoryManagerPlugin") -> None:
+        plugin.manager = self
+        self.plugins.add(plugin)
+
+    async def start(self, scheduler: SchedulerState):
+        """SchedulerPlugin hook. Start kicking off self.run every <interval> seconds
+        until scheduler shutdown.
+        """
+        self.scheduler = scheduler
+
+        async def _():
+            while True:
+                await self.run()
+                await asyncio.sleep(self.interval)
+
+        self.run_task = asyncio.create_task(_())
+
+    async def close(self):
+        """SchedulerPlugin hook. Stop kicking off self.run."""
+        self.run_task.cancel()
+
+    async def run(self):
+        """Run all plugins once and enact their recommendations to replicate/drop keys"""
+        self.workers_memory = {
+            w: w.memory.optimistic for w in self.scheduler.workers.values()
+        }
+        self.pending = defaultdict(lambda: (set(), set()))
+        self._run_plugins()  # populate self.pending
+
+        # TODO WIP act on self.pending
```

Comment:
in scope for this PR; i'm going to use Scheduler.rpc. It will then be moved to bulk comms at a later stage.